pkg/environment/api/api.go

```diff
@@ -73,22 +110,298 @@ func NewEnvironmentService(
 		accountClient:      ac,
 		mysqlClient:        mysqlClient,
 		projectStorage:     v2.NewProjectStorage(mysqlClient),
 		orgStorage:         v2.NewOrganizationStorage(mysqlClient),
 		environmentStorage: v2.NewEnvironmentStorage(mysqlClient),
 		publisher:          publisher,
-		opts:               dopts,
-		logger:             dopts.logger.Named("api"),
+		googleAuthenticator: google.NewAuthenticator(
+			&config.GoogleConfig, logger,
+		),
+		issuer:   issuer,
+		audience: audience,
+		signer:   signer,
+		verifier: verifier,
+		opts:     dopts,
+		logger:   logger,
 	}
 }
 
 func (s *EnvironmentService) Register(server *grpc.Server) {
 	environmentproto.RegisterEnvironmentServiceServer(server, s)
 }
 
+func (s *EnvironmentService) ExchangeDemoToken(
+	ctx context.Context,
+	req *environmentproto.ExchangeDemoTokenRequest,
+) (*environmentproto.ExchangeDemoTokenResponse, error) {
+	localizer := locale.NewLocalizer(ctx)
+	if !s.opts.isDemoSiteEnabled {
+		s.logger.Error("Demo site is not enabled",
+			zap.Any("type", req.Type),
+			zap.String("code", req.Code),
+			zap.String("redirect_url", req.RedirectUrl),
+		)
+		dt, err := statusDemoSiteNotEnabled.WithDetails(&errdetails.LocalizedMessage{
+			Locale:  localizer.GetLocale(),
+			Message: localizer.MustLocalize(locale.Organization),
+		})
+		if err != nil {
+			return nil, auth.StatusInternal.Err()
+		}
+		return nil, dt.Err()
+	}
+	if err := validateExchangeDemoTokenRequest(req, localizer); err != nil {
+		s.logger.Error("Failed to validate the exchange demo token request",
+			zap.Error(err),
+			zap.Any("type", req.Type),
+			zap.String("code", req.Code),
+			zap.String("redirect_url", req.RedirectUrl),
+		)
+		return nil, err
+	}
+	authenticator, err := s.getAuthenticator(req.Type, localizer)
+	if err != nil {
+		s.logger.Error("Failed to get the authenticator",
+			zap.Error(err),
+			zap.Any("type", req.Type),
+			zap.String("code", req.Code),
+			zap.String("redirect_url", req.RedirectUrl),
+		)
+		dt, err := auth.StatusInternal.WithDetails(&errdetails.LocalizedMessage{
+			Locale:  localizer.GetLocale(),
+			Message: localizer.MustLocalize(locale.InternalServerError),
+		})
+		if err != nil {
+			return nil, auth.StatusInternal.Err()
+		}
+		return nil, dt.Err()
+	}
+	userInfo, err := authenticator.Exchange(ctx, req.Code, req.RedirectUrl)
+	if err != nil {
+		s.logger.Error("Failed to exchange",
+			zap.Error(err),
+			zap.Any("type", req.Type),
+			zap.String("code", req.Code),
+			zap.String("redirect_url", req.RedirectUrl),
+		)
+		dt, err := auth.StatusInternal.WithDetails(&errdetails.LocalizedMessage{
+			Locale:  localizer.GetLocale(),
+			Message: localizer.MustLocalize(locale.InternalServerError),
+		})
+		if err != nil {
+			return nil, auth.StatusInternal.Err()
+		}
+		return nil, dt.Err()
+	}
+
+	// check account exist in any organization
+	account, err := s.accountClient.GetMyOrganizationsByEmail(ctx, &accproto.GetMyOrganizationsByEmailRequest{
```

Comment:
This request doesn't include archived organizations.
It would be safer if we use the GetAccountV2 API to ensure the user doesn't exist regardless of the organization status.
Otherwise, if the organization is archived, the user would be able to create another organization.